[Finishes #101455448] Small footer inside app for shared applab apps

Bjvanminnen opened the pull request:
Makes small footer into a react component. Right now this lives in dashboard, and is stored on the dashboard namespace. Long term I think the better approach would be to have shared build a SmallFooter.js bundle, then is then asynchronously loaded as needed by dashboard or apps, but a few things would need to happen to enable that (shared understanding JSX, shared creating multiple bundles, better support for loading JS async, i.e. using partition-bundle)

(note: for some reason licecap is showing some weirdness in that after hiding the menu it has a slightly different gray. same with when i move the cursor. i don't see this behavior outside of licecap)
![footer](https://cloud.githubusercontent.com/assets/1767466/9530308/dede69ec-4cb3-11e5-9fcc-f30614e81582.gif)

For the applab sharepage, we insert a slightly different version of the small footer into our app. This happens on desktop and mobile.
![footershared](https://cloud.githubusercontent.com/assets/1767466/9530402/66aba3b2-4cb4-11e5-8c15-319cd722c53e.gif)

For existing projects, the bottom 34 pixels of space will be covered by the footer. For new projects, prevent this by giving users 34 pixels less of space to work with.

Hamms commented on apps/src/applab/applab.js:
```diff
@@ -103,6 +106,18 @@ function loadLevel() {
+  if (Applab.appHeight === 480) {
+    Applab.footerlessAppHeight = Applab.appHeight - FOOTER_HEIGHT;
+  } else {
+    Applab.footerlessAppHeight = Applab.appHeight;
```
Won't this have unintended behavior when Applab.appHeight is between 480 and (480-FOOTER_HEIGHT)?

Bjvanminnen replied:
Yeah, I think you're right. That will likely never happen, but it should probably be
```
if (Applab.appHeight + FOOTER_HEIGHT >= 480) {
  footerLessAppHeight = Applab.appHeight - FOOTER_HEIGHT
```

davidsbailey replied:
I am concerned that hard-coding 480 here will limit an otherwise flexible thing. AFAIK, today we can create a new applab level with any dimensions and it will work. after this change I'm not sure what the behavior will be.

Bjvanminnen replied:
It's not clear what we should do in terms of the footer for heights > 480. Rather than have unexpected behavior for, I decided to explicitly disallow it. If we decide we need > 480, we'll get an exception and then think about what should happen in terms of the footer in this scenario.

davidsbailey commented:
Is there an end vision for what the mobile/share experience will look like? Specifically, does the decision to take away 34px of app real estate take into consideration the fact that the code/design and view data buttons will be going away?

Bjvanminnen commented:
Yeah, ultimately I believe the buttons above/below the app will go away. There's an item tracking this - I'm not sure if that will be in this sprint or the next.

davidsbailey commented:
asked a different way: Can you explain why the footer needs to go inside the 320x480 instead of outside / below it? Putting it outside of the existing frame seems like it would simplify things a lot.

Bjvanminnen commented:
I know I explored putting it outside of the frame to start with, and ran into some challenges. I don't remember exactly what those were. I could explore that approach again, as it is possible it would simplify a lot of things.

Hamms commented on apps/src/applab/applab.js:
```diff
@@ -292,14 +307,70 @@ function adjustAppSizeStyles(container) {
+      link: 'https://code.org/privacy'
+    }
+  ];
+  if (!dom.isMobile()) {
```
We want to remove the make my own app option `if (dom.isMobile())`, right? This is doing the opposite.

Bjvanminnen replied:
*facepalm* good catch

Hamms commented:
Particularly given that there's nothing in the footer (to my knowledge) that directly relates to the contents of the frame, I much prefer the idea of the footer going outside.

davidsbailey commented on apps/src/applab/applab.js:
```diff
@@ -292,14 +307,70 @@ function adjustAppSizeStyles(container) {
+  var divApplab = document.getElementById('divApplab');
+  var footerDiv = document.createElement('div');
+  footerDiv.setAttribute('id', 'footerDiv');
+  divApplab.parentNode.insertBefore(footerDiv, divApplab.nextSibling);
```
It looks like you are already inserting this as a sibling (rather than a child) of divApplab. So maybe you can keep rendering it this way and just change the math back so that the divApplab element is still 480px tall?

Bjvanminnen replied:
Even if the footer is outside of the divApplab, I believe we'll want divApplab to be 446 (so that when we remove other superfluous buttons a 480 pixel screen consists of the divApplab with the small footer on the bottom.

Bjvanminnen replied:
Actually yeah, right now the footer is outside the existing frame?

Hamms commented on dashboard/app/assets/javascripts/small_footer.js.jsx:
```diff
@@ -0,0 +1,228 @@
+        this.setState({
+          menuState: MenuState.MINIMIZING,
+          moreOffset: 0,
+          hidRecently: true,
```
Does `hidRecently` get used anywhere?

Bjvanminnen replied:
Nope. This was factored out. good catch.

Hamms commented on dashboard/app/assets/javascripts/small_footer.js.jsx:
```diff
@@ -0,0 +1,228 @@
+      }
+
+      if (this.state.menuState === MenuState.COPYRIGHT) {
+        this.setState({ menuState: MenuState.MINIZED });
```
`s/MINIZED/MINIMIZED`

Hamms commented on dashboard/app/assets/javascripts/small_footer.js.jsx:
```diff
@@ -0,0 +1,228 @@
+      }
+
+      if (this.state.menuState === MenuState.EXPANDED) {
+        this.setState({ menuState: MenuState.MINIZED });
```
s/MINIZED/MINIMIZED

davidsbailey commented on dashboard/app/views/layouts/_small_footer.html.haml:
```diff
@@ -1,24 +1,32 @@
-  %li= link_to I18n.t('footer.translate'), 'https://code.org/translate'
-  %li= link_to I18n.t('footer.tos'), 'https://code.org/tos'
-  %li= link_to I18n.t('footer.privacy'), 'https://code.org/privacy'
+:javascript
```
Per our new effort, this code should go in a new javascript file. It should be possible to follow this example pretty closely: https://github.com/code-dot-org/code-dot-org/pull/3491/files

If I understand how the rendering works, you would (for example) name the new file `_small_footer_helper.js` and then call `#{render partial: 'layouts/small_footer_helper', formats: [:js]}` from ruby.

laurelfan replied:
It's not necessary to make it a partial, unless you actually need to render
it inline. You can make it a js file and include it via script tag. See:

https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/views/levels/editors/_all.html.haml
https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/assets/javascripts/editor/blockly_editor.js.erb


On Thu, Aug 27, 2015 at 12:47 PM, davidsbailey <notifications@github.com>
wrote:

> In dashboard/app/views/layouts/_small_footer.html.haml
> <https://github.com/code-dot-org/code-dot-org/pull/3744#discussion_r38138101>
> :
>
> >
> > -%ul#more-menu
> > -  %li= link_to I18n.t('footer.support'), 'https://support.code.org'
> > -  %li= link_to I18n.t('footer.translate'), 'https://code.org/translate'
> > -  %li= link_to I18n.t('footer.tos'), 'https://code.org/tos'
> > -  %li= link_to I18n.t('footer.privacy'), 'https://code.org/privacy'
> > +:javascript
>
> Per our new effort, this code should go in a new javascript file. It
> should be possible to follow this example pretty closely:
> https://github.com/code-dot-org/code-dot-org/pull/3491/files
>
> If I understand how the rendering works, you would (for example) name the
> new file _small_footer_helper.js and then call #{render partial:
> 'layouts/small_footer_helper', formats: [:js]} from ruby.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/code-dot-org/code-dot-org/pull/3744/files#r38138101>.
>

davidsbailey replied:
@laurelfan , Brent will still need to include some inline javascript to pass in the I18n strings in some kind of wrapper object. I think that means he needs a partial, right?

laurelfan replied:
argh, I18n. If you use a `.js.erb` instead of just `.js` you can use I18n.t
in the "js" file.

On Thu, Aug 27, 2015 at 1:23 PM, davidsbailey <notifications@github.com>
wrote:

> In dashboard/app/views/layouts/_small_footer.html.haml
> <https://github.com/code-dot-org/code-dot-org/pull/3744#discussion_r38141764>
> :
>
> >
> > -%ul#more-menu
> > -  %li= link_to I18n.t('footer.support'), 'https://support.code.org'
> > -  %li= link_to I18n.t('footer.translate'), 'https://code.org/translate'
> > -  %li= link_to I18n.t('footer.tos'), 'https://code.org/tos'
> > -  %li= link_to I18n.t('footer.privacy'), 'https://code.org/privacy'
> > +:javascript
>
> @laurelfan <https://github.com/laurelfan> , Brent will still need to
> include some inline javascript to pass in the I18n strings in some kind of
> wrapper object. I think that means he needs a partial, right?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/code-dot-org/code-dot-org/pull/3744/files#r38141764>.
>

Bjvanminnen replied:
I also use view_options. Will this be accessible from js.erb?
```
var i18nDropdown = "#{URI.encode(i18n_dropdown.to_str) if view_options[:has_i18n]}";
```

laurelfan replied:
nope, but I think everything in the ruby view_options is in the js app_options, so you an use that if you can refactor that code to be js.

davidsbailey replied:
It seems like extracting into a .js file (rather than .js.erb) gets us closer to our goals of (1) linting and (2) eventually caching in a CDN.

Bjvanminnen replied:
I'm also using a method from locale_helper (i18n_dropdown). Is there any way to get at that from an erb file?

I wonder if we're ever okay with JS in .haml files when it's mostly config.

Bjvanminnen replied:
I could have a .js file that exports a function createSmallerFooter that takes all of our input as config. Then we still have some JS in our .haml file that extracts the config (strings, i18n_dropdown) from dashboard, and calls the JS method

davidsbailey replied:
OK, good point -- since all you're doing here is passing parameters to a react component, I could believe that it's ok to use inline JS in this case. Furthermore, it is hard to think of any error you could make in this JS that wouldn't be caught by the PropTypes check in the React component. Perhaps we should be lenient in this way only when instantiating React components, to encourage people to use React :-)

davidsbailey commented:
In response to inline comments about reducing the app height, it seems like we should be very deliberate if we're going to change the size of the app window since it will affect users every time we change it. Here are some common device width/height ratios: http://mydevice.io/devices/
applab 320x480 --> 1.5
iphone 6: 375x667 --> 1.78
Samsung S5: 640x360 --> 1.78
many other devices on that list have the same ratio of 1.78, however there are some (including Windows phones and iPhone 4 and older) which have a ratio of 1.5.  So I suppose some questions are (1) would showing the footer outside the frame push the Run/Reset buttons out of view on 1.78 devices? (2) how important is it that we fit the footer into the view (without making the screen unnecessarily narrow) on older phone models?

On the other hand, taking away a few px now (before classes start) and then adding them back later will be easier than leaving them there now and taking them away later. Also, the change you are proposing does essentially leave all of these layout questions the same as before you touched anything, which I guess is what you've been saying all along. So, let's go with yes, subtracting the footer height from the app height sounds good to me. thanks for listening :-)

FYI @sfilman for any additional product input on reducing the default app height.

davidsbailey commented on dashboard/app/assets/javascripts/small_footer.js.jsx:
```diff
@@ -0,0 +1,228 @@
+
+    minimizeOnClickAnywhere: function (event) {
+      // The first time we click anywhere, hide any open children
+      $(document.body).one('click', function (event) {
```
diabolical test case to please check manually: 
(1) run your program
(2) open the footer
(3) hover over a block in the toolbox, or type the partial name of a command in text mode
(4) click "see examples"
(5) click somewhere else

EXPECTED: the footer menu goes away after the second click

I ask because the "see examples" click calls stop propagation. I realize this is a bit of a long shot, but it would suck if there were a circumstance where the footer stayed open forever.

Bjvanminnen replied:
works as expected :)

davidsbailey commented:
Looks good so far. I would like to take another quick look if/after the JS file is extracted.

davidsbailey commented on apps/src/applab/applab.js:
```diff
@@ -231,7 +246,7 @@ function adjustAppSizeStyles(container) {
               // For this scale factor...
               // set the max-height and max-width for the visualization
               childRules[k].style.cssText = "max-height: " +
-                  Applab.appHeight * scale + "px; max-width: " +
```
based on the comments above, I think we want these css height "breakpoints" to stay as the original 480px (Applab.appHeight). This should keep scaling behavior the same as before this change, which seems like what we want if the footer is sitting "inside" the 480px boundary. (here and below)

Bjvanminnen replied:
This scaling only happens on the non-share page. We only end up with the full 480 on the share page, where we add our 34px of footer. On the non-share page we're showing the footerless version of applab (i.e. at 446), hence our use of the footerless height.

davidsbailey commented:
LGTM (and/or please ping me if you want another look)

Hamms commented:
Likewise, pending decision about footer location relative to frame

Bjvanminnen commented on dashboard/app/views/layouts/_small_footer.html.haml:
```diff
@@ -1,24 +1,40 @@
+  };
+
+  window.dashboard.footer.render(React, reactProps,
+    document.getElementById('page-small-footer'));
```
This latest changes moves a little bit more of the JS out (i.e. no React.createElement/React.render). Is there a clean way of having large blocks of ruby inside haml? If so, we could make this even less javascripty by creating our props hash in ruby and passing that to footer.render, i.e. something like
```
#page-small-footer
:ruby
 props = {
   ...
}

:javascript
  window.dashboard.footer.render(React, #{props}, document.getElementById('page-small-footer'))

davidsbailey replied:
yes, you can use inline ruby exactly as you describe. Here is an example: https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/views/levels/_dialog.html.haml#L75

Bjvanminnen commented:
Did some somewhat significant refactoring.
- copyrightStrings get passed to apps via appState instead of living on window.dashboard
- refactored the component itself, in terms of how it deals with menus opening/minimizing properly
- refactored small_footer.js.jsx some to export a render method. done in a way that it should still be easy to extract our SmallFooter component into shared at a future point
- added a hack to deal with the fact that we have two copies of React

davidsbailey commented on apps/test/solutions/applab/ec_screens.js:
```diff
@@ -198,7 +198,7 @@ module.exports = {
         'button("my_button", "my_button_text");' +
         'image("my_image", "http://code.org/images/logo.png");' +
-        'createCanvas("my_canvas", 320, 480);' +
+        'createCanvas("my_canvas", 320, 446);' +
```
446 --> 470, here and below, due to footer height now being 30px

davidsbailey commented:
LGTM. Nice cleanup!
